### PR TITLE
Drag 'n drop

### PR DESCRIPTION
**Pull request by jelly**
@tomasmatus interested?

One tricky thing here is we already have an upload-button component, so the eventlistener should be set up there instead so we don't have to pass `files` around.

---

# Drag-and-Drop file upload

It is now possible to upload files by simply dragging them from your desktop or a file explorer and dropping them directly into Cockpit Files.

### CODE REVIEW

**Comment by jelly** [comment]
This works now, some pain points:

* We should disable dropping when an upload is in progress
* Chrome at least shows that one can drop *everywhere* which is not the case

Test on:

* [ ] Firefox
* [ ] Webkit

**Comment by tomasmatus** [comment]
ok so I think that broke uploading when it should work

**Comment by tomasmatus** [comment]
oh right, tests won't work because I made some local changes to `pkg/lib`.

Also there's a bug I am currently trying to debug - drag&drop doesnt upload to `pwd` but to some old path? (maybe some old state/props, idk, debugging)

**Comment by tomasmatus** [comment]
@garrett I think we can now start talking about how to show that files can be dropped for upload. Right now I have this:

image

Note that it is rather broken in grid view and doesn't work in empty dir so take it as a concept screenshot.

The upload icon appears when user is dragging some file over the files card. Maybe it should instead work on the whole page.
You said you have some ideas how we could design this so I'd appreciate if we could discuss this next week.

File: src/files-folder-view.tsx [hunk 1]
```diff
@@ -26,6 +26,20 @@ import type { FolderFileInfo } from "./app.tsx";
...
+export interface UploadedFilesType {[name: string]:{file: File, progress: number, cancel:() => void}}
+
+interface UploadContextType {
+    uploadedFiles: UploadedFilesType,
+    setUploadedFiles: React.Dispatch<React.SetStateAction<UploadedFilesType>>,
+}
+
+export const UploadContext = createContext({
+    uploadedFiles: {},
+    setUploadedFiles: () => console.warn("UploadContext not initialized!"),
```
Comment (cockpituous) [inline src/files-folder-view.tsx:38]: This added line is not executed by any test.

File: src/upload-button.tsx [hunk 2]
```diff
@@ -258,7 +258,25 @@ export const UploadButton = ({
...
+        cwdInfo?.entries,
+        dialogs,
+        path,
+        setUploadedFiles,
+    ]);
+
+    useEffect(() => {
+        const handleFilesDrop = ((event: CustomEvent) => {
+            onUpload(event.detail);
+        }) as EventListener;
```
Comment (cockpituous) [inline src/upload-button.tsx:272]: These 3 added lines are not executed by any test.

File: src/files-card-body.tsx [hunk 3]
```diff
@@ -273,6 +282,54 @@ export const FilesCardBody = ({
...
+            }
+        };
+
+        const handleDrop = (event: DragEvent) => {
+            event.preventDefault();
+            event.stopPropagation();
+
+            // disable drag & drop when upload is in progress
+            if (isUploading) {
+                return;
```
Comment (cockpituous) [inline src/files-card-body.tsx:316]: These 2 added lines are not executed by any test.

File: src/files-card-body.tsx [hunk 4]
```diff
@@ -273,6 +282,54 @@ export const FilesCardBody = ({
...
+            setDragDropActive(false);
+            dragDropCnt = 0;
+            if (folderViewElem !== null) {
+                folderViewElem.classList.remove("files-drag-hover");
+            }
+        };
+
+        const handleDragOver = (event: DragEvent) => {
+            event.preventDefault();
+            event.stopPropagation();
```
Comment (cockpituous) [inline src/files-card-body.tsx:330]: These 3 added lines are not executed by any test.

File: src/files-folder-view.tsx (hunk 1 above)
Comment (cockpituous) [inline src/files-folder-view.tsx:38]: This added line is not executed by any test.

**Comment by garrett** [comment]
So this is a quick attempt to mock up something.

Notes:

1. Not pixel perfect :wink: ... pixels will be off; sizes are probably off too; the font size is just some arbitrary one in Inkscape, not a PF size (and we should probably use a PF size)
2. Large drop area, over the folder view, so it's a big, easy to achieve drop target
3. 80% opaque background to provide a background contrast to the upload text, but transparent enough to show the files are still there
4. Directory path at the top and summary at the bottom are not obstructed, making it possible to identify where you're dropping it and that there are other files in that directory
5. Since the directory path and the status are not a drop target, they're a fine place to hover over to cancel the drop operation
6. Canceling the drop operation (not dropping on the target) should not load the file; it should prevent the default of browser dragging and dropping from loading the file... ideally this cancelling would also work when dropping on the shell, but I know there are probably technical reasons why this would be difficult
7. Singular vs. plural file/files depending on what's in the hover stack
8. This mockup doesn't show the cursor; I don't know if that's handled by default
9. The outline is dashed 3px PF link color
10. The overlay should have the same size as the card, so it probably needs to be a sibling that comes directly after the card's content and either have the same grid position... or be positioned absolutely.

I'll probably need to help with the CSS.

drop-file-to-upload

WDYT?

**Comment by tomasmatus** [comment]
I think that looks good, I like it! Thanks for the mockup

File: src/files-folder-view.tsx [hunk 5]
```diff
@@ -17,15 +17,35 @@
...
```
Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

File: src/files-folder-view.tsx [hunk 6]
```diff
@@ -57,39 +81,123 @@ export const FilesFolderView = ({
...
+            }
+        };
+
+        const handleDrop = (event: DragEvent) => {
+            event.preventDefault();
+            event.stopPropagation();
+
+            // disable drag & drop when upload is in progress
+            if (isUploading) {
+                return;
```
Comment (cockpituous) [inline src/files-folder-view.tsx:124]: These 2 added lines are not executed by any test.

File: src/files-folder-view.tsx [hunk 7]
```diff
@@ -57,39 +81,123 @@ export const FilesFolderView = ({
...
+
+            cockpit.assert(event.dataTransfer !== null, "dataTransfer cannot be null");
+            dispatchEvent(new CustomEvent('files-drop', { detail: event.dataTransfer.files }));
+            setDragDropActive(false);
+            dragDropCnt = 0;
+        };
+
+        const handleDragOver = (event: DragEvent) => {
+            event.preventDefault();
+            event.stopPropagation();
```
Comment (cockpituous) [inline src/files-folder-view.tsx:135]: These 3 added lines are not executed by any test.

File: src/files-folder-view.tsx (hunk 5 above)
Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

File: src/files-folder-view.tsx [hunk 8]
```diff
@@ -57,39 +81,121 @@ export const FilesFolderView = ({
...
+            }
+        };
+
+        const handleDrop = (event: DragEvent) => {
+            event.preventDefault();
+            event.stopPropagation();
+
+            // disable drag & drop when upload is in progress
+            if (isUploading) {
+                return;
```
Comment (cockpituous) [inline src/files-folder-view.tsx:122]: These 2 added lines are not executed by any test.

**Comment by tomasmatus** [comment]
> 7. Singular vs. plural file/files depending on what's in the hover stack

I'm afraid this is not possible, the `DataTransfer` object is added only on `dragdrop` event, for `dragenter`, `draghover` or `dragleave` it's set to `null`.

File: src/files-folder-view.tsx (hunk 5 above)
Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

File: src/files-folder-view.tsx (hunk 8 above)
Comment (cockpituous) [inline src/files-folder-view.tsx:122]: These 2 added lines are not executed by any test.

File: src/files-folder-view.tsx (hunk 5 above)
Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

File: test/check-application [hunk 9]
```diff
@@ -2204,6 +2204,90 @@ class TestFiles(testlib.MachineCase):
...
+                      }''')
+
+            def drag_file_event(selector: str, dragType: str, filename: str | None = None) -> None:
+                b.wait_visible(selector)
+                b.call_js_func('dragFileEvent', selector, dragType, filename)
+
+            b.wait_not_present(".drag-drop-upload")
+            drag_file_event(".fileview-wrapper", 'dragenter')
+            b.wait_visible(".drag-drop-upload")
+            # TODO add pixeltest here when we have a design what the UI should look like during drag events
```
Comment (jelly) [inline test/check-application:2240]: That seems now :)

File: src/files-card-body.tsx [hunk 10]
```diff
@@ -155,10 +157,12 @@ export const FilesCardBody = ({
     clipboard: string[], setClipboard: React.Dispatch<React.SetStateAction<string[]>>,
     showHidden: boolean,
     setShowHidden: React.Dispatch<React.SetStateAction<boolean>>,
+    setDragDropActive: React.Dispatch<React.SetStateAction<boolean>>,
 }) => {
     const [boxPerRow, setBoxPerRow] = useState(0);
     const dialogs = useDialogs();
     const { addAlert, cwdInfo } = useFilesContext();
+    const { uploadedFiles } = useContext(UploadContext);
```
Comment (jelly) [inline src/files-card-body.tsx:165]: Why? It is not used here? Nor is anything else relevant to Upload, it just gets passed down.

File: src/files-folder-view.tsx [hunk 11]
```diff
@@ -17,15 +17,35 @@
...
+export interface UploadedFilesType {[name: string]:{file: File, progress: number, cancel:() => void}}
+
+interface UploadContextType {
+    uploadedFiles: UploadedFilesType,
+    setUploadedFiles: React.Dispatch<React.SetStateAction<UploadedFilesType>>,
+}
+
+export const UploadContext = createContext({
+    uploadedFiles: {},
+    setUploadedFiles: () => console.warn("UploadContext not initialized!"),
```
Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

Comment (cockpituous) [inline src/files-folder-view.tsx:44]: This added line is not executed by any test.

**Review by tomasmatus — APPROVED** [review]
Looks good, thanks! 🚀

**Pull request merged**